1\. **Pandas DataFrame**

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a pair of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every `x` BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [2]:
# If haven't downloaded it yet, please get the data file with wget
!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ./data/

--2022-11-15 19:09:23--  https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt
Resolving www.dropbox.com (www.dropbox.com)... 162.125.69.18
Connecting to www.dropbox.com (www.dropbox.com)|162.125.69.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: /s/raw/xvjzaxzz3ysphme/data_000637.txt [following]
--2022-11-15 19:09:23--  https://www.dropbox.com/s/raw/xvjzaxzz3ysphme/data_000637.txt
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc975137460173d4b6e775ed0834.dl.dropboxusercontent.com/cd/0/inline/Bw2zocf1BSYsgEFcfhFNJOIE4m3609c3b4pqZDrcOK9e-ZRhd6MdmavupPIHtaoWV-s7Z9c3iLNPe3pRXqOKpttyFOKf9Tqqnxrtup9SStOWnYHmPOnbUK5EuS74Oa7VsfK9b5O1tN4rWzgp-_H00rUxcCSIbFTgnCES6QEUs-oYaw/file# [following]
--2022-11-15 19:09:24--  https://uc975137460173d4b6e775ed0834.dl.dropboxusercontent.com/cd/0/inline/Bw2zocf1BSYsgEFcfhFNJOIE4m3609c3b4pqZDrcOK9e-ZRhd6MdmavupPIHtaoWV-s7Z9c3iLNPe3pRXqOKpttyFOKf9Tqqn

1\. Create a Pandas DataFrame reading N rows of the `data/data_000637.txt` dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k (check the documentation).

In [23]:
import pandas as pd

N = 10000
file_name = './data/data_000637.txt'
data = pd.read_csv(file_name)
data = data[:N]
data

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0        1     0          123  3869200167        2374        26
1        1     0          124  3869200167        2374        27
2        1     0           63  3869200167        2553        28
3        1     0           64  3869200167        2558        19
4        1     0           64  3869200167        2760        25
...    ...   ...          ...         ...         ...       ...
9995     1     0           61  3869200267         696        11
9996     1     0           60  3869200267         701         5
9997     1     0           59  3869200267         707        23
9998     1     0           63  3869200267         706        15
9999     1     0           49  3869200267         777        13

[10000 rows x 6 columns]

2\. Estimate the number of BX in a ORBIT (the value `x`).

*Hint*: check when the BX counter reaches the maximum value before being reset to 0.

In [52]:
x = data['BX_COUNTER'].max()
x

3563

3\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information) since the beginning of the data acquisition, and convert the new column to a Time Series.

In [32]:
data['ABSOLUTE_TIME_NS'] = data['ORBIT_CNT']*25*x + data['BX_COUNTER']*25 + data['TDC_MEAS']*(25/30)

tseries = pd.Series(data['ABSOLUTE_TIME_NS'])
tseries

0       3.446490e+14
1       3.446490e+14
2       3.446490e+14
3       3.446490e+14
4       3.446490e+14
            ...     
9995    3.446490e+14
9996    3.446490e+14
9997    3.446490e+14
9998    3.446490e+14
9999    3.446490e+14
Name: ABSOLUTE_TIME_NS, Length: 10000, dtype: float64

4\. Find out the duration of the data taking in hours, minutes and seconds, by using the features of the Time Series. Perform this check reading the whole dataset.

In [36]:
#ts = pd.to_datetime(tseries, format= '%s')

5\. Use the `.groupby()` method to find out the noisy channels, i.e. the TDC channels with most counts (print to screen the top 3 and the corresponding counts)

In [50]:
noisy_channels = data.groupby('TDC_CHANNEL')['TDC_CHANNEL'].count()
noisy_channels = noisy_channels.sort_values(ascending=False)
noisy_channels.iloc[0:3]

TDC_CHANNEL
139    860
63     499
64     491
Name: TDC_CHANNEL, dtype: int64

6\. Count the number of non-empty orbits (i.e. the number of orbits with at least one hit).

In [ ]:
NE_ORBITS = data.

7\. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139.

8\. Create two Series (one for each FPGA) that have the TDC channel as index, and the number of counts for the corresponding TDC channel as values.

9\. **Optional:** Create two histograms (one for each FPGA) that show the number of counts for each TDC channel.